react-shopping-cart/src/pages/ProductDetailPage/ProductDetailPage.jsx

```diff
@@ -0,0 +1,77 @@
+import { useDispatch, useSelector } from "react-redux";
+import { useParams } from "react-router-dom";
+import ProductName from "../../component/@shared/ProductName/ProductName";
+import ProductPrice from "../../component/@shared/ProductPrice/ProductPrice";
+import ProductThumbnail from "../../component/@shared/ProductThumbnail/ProductThumbnail";
+import {
+  selectDetailProduct,
+  selectProductsLoading,
+} from "../../redux/products/products.selector";
+import { ColumnFlexWrapper, RowFlexWrapper } from "../../styles/Wrapper";
+import {
+  selectCartsLoading,
+  selectCurrentCarts,
+} from "../../redux/carts/carts.selector";
+import { isInCart } from "../../util/check";
+import useClickCartButton from "../../hooks/useClickCartButton";
+import ShoppingCartButton from "../../component/ShoppingCartButton/ShoppingCartButton";
+import { CURRENT_USER } from "../../constants/index";
+import { useEffect } from "react";
+import { fetchProductDetailStart } from "../../redux/products/products.action";
+import WithSpinner from "../../component/@shared/WithSpinner/WithSpinner";
+
+function ProductDetailPage() {
+  const { idx } = useParams();
+  const product = useSelector(selectDetailProduct);
+  const carts = useSelector(selectCurrentCarts);
+  const cartsLoading = useSelector(selectCartsLoading);
+  const productsLoading = useSelector(selectProductsLoading);
+
+  const dispatch = useDispatch();
+  const { handleAddProduct, handleDeleteProduct } = useClickCartButton();
+
+  const isCartItem = isInCart(idx, carts);
+
+  useEffect(() => {
+    dispatch(fetchProductDetailStart(idx));
+  }, [dispatch, idx]);
+
+  return (
+    product && (
+      <WithSpinner loading={cartsLoading || productsLoading}>
+        <ColumnFlexWrapper gap="20px" width="425px" ml="auto" mr="auto">
+          <ProductThumbnail type="detail" src={product.image} />
+          <ProductName type="detail">{product.name}</ProductName>
+          <RowFlexWrapper
+            gap="300px"
+            bt="2px"
+            bColor="gray_01"
+            width="100%"
+            padding="15px 0px"
+          >
+            <div>금액</div>
+            <ProductPrice type="detail">{product.price}원</ProductPrice>
+          </RowFlexWrapper>
+          <ShoppingCartButton
+            $isincart={isCartItem}
+            onClick={
+              isCartItem
+                ? (e) => handleDeleteProduct(e, `${CURRENT_USER}${idx}`)
+                : (e) =>
+                    handleAddProduct(e, {
+                      name: product.name,
+                      price: product.price,
+                      id: idx,
+                      thumbnail: product.image,
+                    })
+            }
+          >
+            {isCartItem ? "장바구니에서 제거" : "장바구니에 추가"}
+          </ShoppingCartButton>
```

Comment:
음 저는 `ProductDetailPage.jsx` 에 있는 `ShoppingCartButton` 관련 로직들이 `ShoppingCartButton` 컴포넌트 내부로 이동이 된 후 아래와 같이 사용이 되면 좋을 것 같은데요... 전반적으로 컴포넌트의 분리나 역할 등이 일관성이 많이 부족한 것 같습니다 ;ㅁ;ㅜㅜ 컴포넌트는 많은데 표현 방법이 다양해서 조금 파악하기 어려운 것 같아요 ㅠ
```suggestion
          <ShoppingCartButton />
```

Reply:
너무 styled로만 하나를 만들고 로직들을 그밖에서 하려고 했던것 같네요. react를 좀더 선언적으로 작성을 해야할 필요가 느껴지네요..🥲